File: python/tests/power/ade7953_test_linux.py
"""ADE7953 hardware-in-loop test for Linux (smbus2)."""

import os
import time
from periph.connection.i2c_linux import I2CConnection
from periph.chips.power.ade7953 import ADE7953Full, ADE7953Source

passed = 0
failed = 0


def check_eq(label, got, expected):
    global passed, failed
    if got == expected:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL {}: got {!r}, expected {!r}'.format(label, got, expected))
        failed += 1


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


I2C_BUS = int(os.environ.get('LINUX_I2C_BUS', '1'))
I2C_ADDR = int(os.environ.get('I2C_ADDR', '0x38'), 16)

VOLTAGE_GAIN = 251.0
CURRENT_GAIN = 30.0

connection = I2CConnection(I2C_BUS, I2C_ADDR)
try:
    ade = ADE7953Full(connection, VOLTAGE_GAIN, CURRENT_GAIN)

    time.sleep(0.2)

    check_true('voltage non-negative', ade.voltage() >= 0.0)
    check_true('current non-negative', ade.current() >= 0.0)
    check_true('active_power finite', ade.active_power() > -1000000.0)
    check_true('active_energy finite', ade.active_energy() > -1000.0)

    ade.enable_interrupt(ADE7953Source.ZXV)
    check_true('IRQSTATA readable', ade.interrupt_status('a') >= 0)
    ade.clear_interrupts('a')
    ade.disable_interrupt(ADE7953Source.ZXV)

    ade.configure_no_load(active=0x1234)
    ade.disable_no_load(active=True)

    ade.reset()
    check_true('voltage after reset', ade.voltage() >= 0.0)

    print('===DONE: {} passed, {} failed==='.format(passed, failed))
finally:
    connection.close()

File: cpp/tests/power/ade7953_test_linux/ade7953_test_linux.cpp
#include <cstdio>
#include <cstdint>
#include "I2CConnectionLinux.h"
#include "ADE7953.h"

#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x38
#endif

static const float VOLTAGE_GAIN = 251.0f;
static const float CURRENT_GAIN = 30.0f;

static int passed = 0;
static int failed = 0;

static void check_true(const char* label, bool cond) {
    if (cond) { std::printf("PASS %s\n", label); passed++; }
    else      { std::printf("FAIL %s\n", label); failed++; }
}

int main() {
    I2CConnectionLinux conn(TEST_I2C_BUS, TEST_ADDR);
    ADE7953Full ade(conn, VOLTAGE_GAIN, CURRENT_GAIN);

    check_true("voltage non-negative", ade.voltage() >= 0.0f);
    check_true("current non-negative", ade.current() >= 0.0f);
    check_true("activePower finite",   ade.activePower() > -1.0e6f);
    check_true("activeEnergy finite",  ade.activeEnergy() > -1000.0f);

    check_true("linePeriod positive",  ade.linePeriod() > 0.0f);

    ade.reset();
    check_true("voltage after reset", ade.voltage() >= 0.0f);

    std::printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}